blessed/sequences.py

```diff
@@ -816,3 +820,73 @@ def horizontal_distance(ucs, term):
     return (termcap_distance(ucs, 'cub', -1, term) or
             termcap_distance(ucs, 'cuf', 1, term) or
             0)
+
+
+def identify_fragment(term, ucs):
+    # simple terminal control characters,
+    ctrl_seqs = u'\a\b\r\n\x0e\x0f'
+
+    if any([ucs.startswith(_ch) for _ch in ctrl_seqs]):
+        return TextFragment(ucs[:1], True, '?', None)
+
+    ret = term and (
+        regex_or_match(term._will_move, ucs) or
+        regex_or_match(term._wont_move, ucs)
+    )
+    if ret:
+        return ret
+
+    # known multibyte sequences,
+    matching_seq = term and (
+        term._re_cub and term._re_cub.match(ucs) or
+        term._re_cuf and term._re_cuf.match(ucs)
+    )
+
+    if matching_seq:
+        return TextFragment(matching_seq.group(), True, '?', None)
+
+    raise ValueError("not a sequence!: "+repr(ucs))
+
+
+def regex_or_match(patterns, ucs):
+    for cap, pat in patterns:
+        matching_seq = re.match(pat, ucs)
+        if matching_seq is None:
+            continue
+        params = matching_seq.groups()[1:]
+        return TextFragment(matching_seq.group(), True, cap, params if params else None)
+    return None
+
+def i_and_split_nonsequences(fragments):
+    """Yields simple unicode characters or sequences with index"""
+    idx = 0
+    for part in fragments:
+        if part.is_sequence:
+            yield idx, part
+            idx += len(part.ucs)
+        else:
+            for c in part.ucs:
+                yield idx, TextFragment(c, False, None, None)
+                idx += 1
+
+def iter_parse(term, ucs):
+    r"""
+
+
+    """
+    outp = u''
+    nxt = 0
+    for idx in range(0, six.text_type.__len__(ucs)):
+        if idx == nxt:
+            l = measure_length(ucs[idx:], term)
+            if l == 0:
+                outp += ucs[idx]
+                nxt += 1
+            else:
+                if outp:
+                    yield TextFragment(outp, False, None, None)
+                    outp = u''
+                yield identify_fragment(term, ucs[idx:idx+l])
+                nxt += l
+    if outp:
+        yield TextFragment(outp, False, None, None)
```

Comment:
The hope is by reducing to a common function, we can find ways to optimize that function, feel free to innovate